### PR TITLE
Adding new cards to contest view

### PR DESCRIPTION
**Pull request by akuleshov7**
### What's done:
- new cards at contest view

image

### CODE REVIEW

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * [Props] of the project rating functional interface
+ */
+external interface ProjectRatingProps : PropsWithChildren {
+    /**
+     * logged-in user or null
+     */
+    var userInfo: UserInfo?
+}
+
+// TODO: investigate why user properties are not propagated into the FC if we refresh the page (F5),
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:27]: @Cheshiriks @sanyavertolet @petertrr I am pretty sure that you have faced this kind of issue already - pls suggest something, if you know it

Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:27]: resolved with useDeferredRequest

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt [hunk 2]
```diff
@@ -46,6 +49,30 @@ class LnkUserProjectController(
     private val projectService: ProjectService,
     private val projectPermissionEvaluator: ProjectPermissionEvaluator,
 ) {
+    @GetMapping(path = ["/current-user"])
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt:52]: Seems a bit unclear, maybe `/get-by-current-user` or just `/by-current-user`?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/ContestsStatistics.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            headers = jsonHeaders,
+            loadingHandler = ::loadingHandler,
+        )
+            .decodeFromJsonString()
+        setFinishedContests(contests.toSet())
+    }
+
+    div {
+        className = ClassName("col-lg-4")
+        ReactHTML.div {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/ContestsStatistics.kt:51]:
```suggestion
        div {
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/YourContests.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import react.dom.html.ReactHTML.div
+
+import kotlinx.js.jso
+
+val yourContests = yourContests()
+
+private fun yourContests() = VFC {
+    div {
+        className = ClassName("col-lg-5")
+        ReactHTML.div {
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/YourContests.kt:20]:
```suggestion
        div {
```

**Review by sanyavertolet — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+import kotlinx.js.jso
+
+val myProjectsRating = myProjectsRatings()
+
+/**
+ * @return functional component
+ */
+fun myProjectsRatings() = FC<ContestListViewProps> { props ->
+    val (myProjects, setMyProjects) = useState<Set<Project>>(emptySet())
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:26]:
```suggestion
    val (myProjects, setMyProjects) = useState(emptySet<Project>())
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    val getMyProjects = useDeferredRequest {
+        setMyProjects(
+            get(
+                url = "$apiUrl/projects/current-user?userId=${props.currentUserInfo?.id}",
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:30]:
```suggestion
                url = "$apiUrl/projects/current-user${props.currentUserInfo?.id?.let { "?userId=$it" } ?: ""}",
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                headers = jsonHeaders,
+                loadingHandler = ::loadingHandler,
+            ).decodeFromJsonString<List<Project>>()
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:33]: will it be a problem decode to Set?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                        }
+                    }
+                    div {
+                        className = ClassName("row")
+                        style = jso {
+                            alignItems = AlignItems.center
+                            justifyContent = JustifyContent.center
+                        }
+                        h4 {
+                            +"432796"
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:73]: This is already available as `Project.contestRating`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/YourContests.kt (hunk 4 above)
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/YourContests.kt:20]: thx

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt (hunk 8 above)
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:73]: in next PR

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt (hunk 6 above)
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:30]: made it a little bit differently - I don't want to have a request with empty string to the backend...

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt (hunk 7 above)
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:33]: nope

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt [hunk 9]
```diff
@@ -46,6 +49,30 @@ class LnkUserProjectController(
...
+        Parameter(name = "projectName", `in` = ParameterIn.PATH, description = "name of a project", required = true),
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully fetched users from project.")
+    @ApiResponse(responseCode = "404", description = "Project with such name was not found.")
+    fun getProjectsOfCurrentUser(@RequestParam userId: Long?, authentication: Authentication): Flux<Project> {
+        val userIdFromAuth = (authentication.details as AuthenticationDetails).id
+        return Flux.fromIterable(
+            if (userId == userIdFromAuth) lnkUserProjectService.getNonDeletedProjectsByUserId(userId) else emptyList()
+        ).switchIfEmptyToNotFound {
+            "You have requested projects for a user that isn't you. Are you a hacker? Please don't"
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt:70]: we won't request - in first condition `if (userId == userIdFromAuth)` we will check that user is valid, after that we will return empty set.

You mean that it should be checked from the frontend?

[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt:70]: Do you even need to pass the userID from the front? Why not immediately return values from the controller using only the Authentication value?

[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt:70]: I have refactored that a little bit

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                url = "$apiUrl/projects/current-user?userId=${props.currentUserInfo?.id}",
+                headers = jsonHeaders,
+                loadingHandler = ::loadingHandler,
+            ).decodeFromJsonString<List<Project>>()
+                .toSet()
+        )
+    }
+
+    props.currentUserInfo?.let {
+        getMyProjects()
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/contests/MyProjectsRating.kt:40]: AH! definitely - if is changing the number of setState and can cause a lot of problems

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt]: > Do you even need to pass the userID from the front? Why not immediately return values from the controller using only the Authentication value?

I suppose it is much better to have extendable method that can be used in future.

**Pull request merged**